packages/@aws-cdk/tmp-toolkit-helpers/src/api/refactoring/index.ts

```diff
@@ -0,0 +1,415 @@
+      template,
+    };
+  };
+
+  // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
+  return Promise.all(summaries.map(normalize));
+}
+
+/**
+ * Computes the digest for each resource in the template.
+ *
+ * Conceptually, the digest is computed as:
+ *
+ *     digest(resource) = hash(type + properties + dependencies.map(d))
+ *
+ * where `hash` is a cryptographic hash function. In other words, the digest of a
+ * resource is computed from its type, its own properties (that is, excluding
+ * properties that refer to other resources), and the digests of each of its
+ * dependencies.
+ *
+ * The digest of a resource, defined recursively this way, remains stable even if
+ * one or more of its dependencies gets renamed. Since the resources in a
+ * CloudFormation template form a directed acyclic graph, this function is
+ * well-defined.
+ */
+export function computeResourceDigests(template: CloudFormationTemplate): Record<string, string> {
+  const resources = template.Resources || {};
+  const graph: Record<string, Set<string>> = {};
+  const reverseGraph: Record<string, Set<string>> = {};
+
+  // 1. Build adjacency lists
+  for (const id of Object.keys(resources)) {
+    graph[id] = new Set();
+    reverseGraph[id] = new Set();
+  }
+
+  // 2. Detect dependencies by searching for Ref/Fn::GetAtt
+  const findDependencies = (value: any): string[] => {
+    if (!value || typeof value !== 'object') return [];
+    if (Array.isArray(value)) {
+      return value.flatMap(findDependencies);
+    }
+    if ('Ref' in value) {
+      return [value.Ref];
+    }
+    if ('Fn::GetAtt' in value) {
+      const refTarget = Array.isArray(value['Fn::GetAtt']) ? value['Fn::GetAtt'][0] : value['Fn::GetAtt'].split('.')[0];
+      return [refTarget];
+    }
+    return Object.values(value).flatMap(findDependencies);
+  };
+
+  for (const [id, res] of Object.entries(resources)) {
+    const deps = findDependencies(res.Properties || {});
+    for (const dep of deps) {
+      if (dep in resources && dep !== id) {
+        graph[id].add(dep);
+        reverseGraph[dep].add(id);
+      }
+    }
+  }
+
+  // 3. Topological sort
+  const inDegree = Object.keys(graph).reduce((acc, k) => {
+    acc[k] = graph[k].size;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const queue = Object.keys(inDegree).filter((k) => inDegree[k] === 0);
+  const order: string[] = [];
+
+  while (queue.length > 0) {
+    const node = queue.shift()!;
+    order.push(node);
+    for (const nxt of reverseGraph[node]) {
+      inDegree[nxt]--;
+      if (inDegree[nxt] === 0) {
+        queue.push(nxt);
+      }
+    }
+  }
+
+  // 4. Compute digests in sorted order
+  const result: Record<string, string> = {};
+  for (const id of order) {
+    const resource = resources[id];
+    const depDigests = Array.from(graph[id]).map((d) => result[d]);
+    const propsWithoutRefs = JSON.stringify(stripReferences(stripConstructPath(resource.Properties)));
```

Comment:
What about other top-level CloudFormation attributes? Like CreationPolicy, Condition, etc? Should we not incorporate those as well?